### PR TITLE
feat: Implement FP32 accumulation for matmul

### PR DESCRIPTION
**Pull request by peri044**
# Description

Implement FP32 accumulation for matmul layers 

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 1]
```diff
@@ -221,7 +225,7 @@ def compile(
     logger.debug("Input graph: " + str(gm.graph))
 
     # Apply lowering on the graph module
-    gm = post_lowering(gm)
+    gm = post_lowering(gm, use_fp32_acc=use_fp32_acc)
```
Comment (HolyWu) [inline py/torch_tensorrt/dynamo/_compiler.py:228]: Should a check be added to check whether only fp16 is the only precision in `enabled_precisions` and raise an error if it's not?

Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py:228]: Yes. I'm planning to add that. This is currently WIP.

File: py/torch_tensorrt/dynamo/lowering/passes/accumulate_fp32_matmul.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import torch
+from torch_tensorrt.dynamo.lowering.passes.pass_utils import (
+    clean_up_graph_after_modifications,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def accumulate_fp32_matmul(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/lowering/passes/accumulate_fp32_matmul.py:11]: tests for this pass?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 3]
```diff
@@ -66,10 +66,11 @@ def __init__(
...
         flag = 0
-
-        # It is deprecated to not use this flag
-        EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
-        flag |= EXPLICIT_BATCH
+        if compilation_settings.use_strong_types:
+            STRONGLY_TYPED = 1 << (int)(
+                trt.NetworkDefinitionCreationFlag.STRONGLY_TYPED
+            )
+            flag |= STRONGLY_TYPED
```
Comment (keehyuna) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:73]: Does this need change in ops like below?
https://github.com/pytorch/TensorRT/pull/3111/files#diff-89deb3709966060e627cf17441fb7d47a72cb12faf484da6c6b968f9c72b2811R37-R55
In my test with sd_unet model, I got error from IConvolutionLayer/INormalizationLayer/IConcatenationLayer when strongly typed network is used. If multiple input is used for layer, trt expected same dtype when strongly typed network is enabled.

File: examples/dynamo/torch_export_gpt2.py [hunk 4]
```diff
@@ -56,6 +60,8 @@
     truncate_double=True,
     device=DEVICE,
     disable_tf32=True,
+    use_strong_types=True,
```
Comment (narendasan) [inline examples/dynamo/torch_export_gpt2.py:63]: Thoughts on calling this `use_strong_typing`?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -197,6 +201,20 @@ def compile(
...
+    if use_strong_types:
+        if len(enabled_precisions) != 1 or not any(
+            x in enabled_precisions for x in {torch.float32, dtype.f32}
+        ):
+            raise AssertionError(
+                f"When use_strong_types is enabled, only torch.float32 is allowed in the enabled_precisions but found {enabled_precisions}"
+            )
+
+    if use_fp32_acc:
+        logger.debug(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:213]: Should there be deeper input validation here? like checking if fp16 is in enabled precisions and throwing a warning if fp16 is not in the set but use_fp32_acc is True?  The second line should be part of the docs.

File: examples/dynamo/torch_export_gpt2.py (hunk 4 above)
Comment (peri044) [reply examples/dynamo/torch_export_gpt2.py:63]: sounds good

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py [hunk 6]
```diff
@@ -13,13 +13,11 @@
...
     cast_trt_tensor,
     get_trt_tensor,
-)
-from torch_tensorrt.fx.converters.converter_utils import (
     broadcast,
     has_dynamic_shape,
     set_layer_name,
 )
-from torch_tensorrt.fx.types import TRTElementWiseOp, TRTTensor
+from torch_tensorrt.dynamo.types import TRTElementWiseOp, TRTTensor
```
[the remark this answers is not in the document]
Comment (peri044) [reply py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py:20]: Got it. Opened an issue https://github.com/pytorch/TensorRT/issues/3222. I will work on this intermittently

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 5 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py:213]: -use_fp32_acc - replicate pytorch fp16 matmul behavior 
- use_strong typing - some layers in llm moels are explicity casted to fp32 (replicate this in TRT) 
- enabled_precisions - should be not have any other than FP32 

- amp example to demo strong typing (userguide/ tutorial) 
- use_fp32_acc + use_strong_typing (llm examples demo)
